### DISCUSSION TITLE
Merging multiple configuration sources

### DISCUSSION THREAD

**Discussion by pinkfloydx33** [dotnet/yarp#1488] [Q&A] [answered]
Is it possible to combine multiple configuration sources? Specifically I'm interested in a case where some parts of the configuration are defined in config/code and the rest is produced out of the k8s controller. I'm imaging an instance where you'd want to include a single set of transforms across all routes, but still use the k8s controller to read from ingresses.

For example, you might have a series of headers that you want added to all routes--something you can't currently set directly on an Ingress but would want to avoid repeating even if you could. I'm also thinking of defining "Exact" Ingress paths with placeholders in them, ie. `path/{plugin}/whatever`, but I'm not sure how to then get YARP to apply the necessary transforms if it's building the config solely based on the ingress definitions.

Or am I thinking about this all wrong and overlooking the correct way of doing this? Do I simply just chain `AddTransforms` after to the call `FromMessages`? Is it that easy?

**Comment by Tratcher** [comment, marked as answer]
If you only want to modify existing routes, then yes AddTransforms is a good way to do that.
https://github.com/microsoft/reverse-proxy/blob/23d235fe9fde53419e8a40c2db02c8bdae972eaf/samples/ReverseProxy.Transforms.Sample/Startup.cs#L39

There are also config filters if you want to tweak the raw model before it's processed.
https://github.com/microsoft/reverse-proxy/blob/23d235fe9fde53419e8a40c2db02c8bdae972eaf/samples/ReverseProxy.ConfigFilter.Sample/CustomConfigFilter.cs#L55

**Reply by pinkfloydx33** [reply]
Cool thanks. I actually started playing with both yesterday. Are AddTransforms and AddConfigFilter guaranteed to execute it in the order in which they are called in source or is there a scenario where they could potentially be executed out of order?

Side note: I couldn't find any documentation on AddConfigFilter/IProxyConfigFilter other than the sample you linked to (which is actually how I stumbled across my answer). Am I overlooking something in the docs or is this actually missing?

**Reply by Tratcher** [reply]
> I actually started playing with both yesterday. Are AddTransforms and AddConfigFilter guaranteed to execute it in the order in which they are called in source or is there a scenario where they could potentially be executed out of order?

Do you mean if you call AddTransforms multiple times, do those always execute in the order registered? Yes. Each gets registered in DI, then the whole list is resolved, and most DI containers preserve that registration ordering. 
https://github.com/microsoft/reverse-proxy/blob/a5b36816f11e939f5037f521109153b5d67bae26/src/ReverseProxy/Management/ReverseProxyServiceCollectionExtensions.cs#L108
https://github.com/microsoft/reverse-proxy/blob/a5b36816f11e939f5037f521109153b5d67bae26/src/ReverseProxy/Management/ReverseProxyServiceCollectionExtensions.cs#L92

> Side note: I couldn't find any documentation on AddConfigFilter/IProxyConfigFilter other than the sample you linked to (which is actually how I stumbled across my answer). Am I overlooking something in the docs or is this actually missing?

You're right, this isn't covered in the docs. I've filed #1492 for that.
